### DISCUSSION TITLE
Custom transform3d handler questions

### DISCUSSION THREAD

**Discussion by csyshing** [Autodesk/maya-usd#1811] [Q&A] [answered]
Hi,

For historical reasons, our studio has some USD assets that have xformOp names with extra suffix, for example:
 - xformOp:translate:translate
 - xformOp:rotateXYZ:rotate
 - xformOp:scale:scale

We used to treat and handle those xformOps as they were no extra suffixes:
 - xformOp:translate:translate  --> xformOp:translate
 - xformOp:rotateXYZ:rotate --> xformOp:rotateXYZ
 - xformOp:scale:scale  --> xformOp:scale

`maya-usd` has the transform3d handler design, but unfortunately [UsdTransform3dMayaXformStack](https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.h) cannot recognize those xformOp with extra suffixes, I was thinking to insert our own xformOp handler to deal with those xformOps, so here are a few questions regarding to the current design:

1. To let `mayaUsdPlugin` and `AL_USDMaya` plugins to pick up my custom handler, the only solution I found so far, is to patch the global registration code in [lib/mayaUsd/ufe/Global.cpp](https://github.com/Autodesk/maya-usd/blob/395ad3177179161378296d85a822a63b5dcc9366/lib/mayaUsd/ufe/Global.cpp#L169-L177), is there any other better way to insert custom xformOp handler without dirtying the shared code base? The problem is that, there is no guarantee which plugin will be loaded first, if I keep my changes in `AL_USDMaya`​ plugin, `mayaUsdPlugin​` might reset the transform3d handler later when it is loaded.

2. My custom handler inherits from [UsdTransform3dMayaXformStack](https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.h), overrides the methods `translation()`, `rotate()`, `scale()`, `translateCmd()`, `rotateCmd()` and `scaleCmd()` so that the custom handler can read/edit the xformOps as expected (as they were no extra suffix). `UsdTransform3dMayaXformStack` provides `UsdVecOpUndoablCmd` and `UsdRotateOpUndoableCmd` classes for UFE undoable support, but they are not in public headers so I ended up need to duplicate [quite some lines from `UsdTransform3dMayaXformStack.cpp`](https://github.com/Autodesk/maya-usd/blob/dev/lib/mayaUsd/ufe/UsdTransform3dMayaXformStack.cpp#L184-L372) into my own handler file, my question is, can we get those code abstracted and moved to a common place to reduce code duplication?

Thanks,
Zhicheng

**Comment by ppt-adsk** [comment, marked as answer]
Hi @csyshing, not sure if I can give a complete answer, but here are some thoughts that I hope you'll find useful:

1. Even if you load multiple plugins, Global.cpp:initialize() should only be called once:
https://github.com/Autodesk/maya-usd/blob/395ad3177179161378296d85a822a63b5dcc9366/lib/mayaUsd/ufe/Global.cpp#L119
My hope is that this would let you place yourself at the head of the Transform3d handler chain of responsibility in the AL plugin (and perhaps in the mayaUsdPlugin if you need that too) in a reliable way.  In the plugin initialization, as long as Global.cpp:initialize() has been called once, then you would retrieve the current Transform3dHandler, which is in fact the head of the chain of responsibility of Transform3d handlers.  You would use that as the argument to your own Transform3d handler, and if your Transform3d handler recognizes your special, backward-compatibility transform ops, it does the work itself, otherwise it just delegates to the existing Transform3d handler.  This places your Transform3d handler at the top of the chain of responsibility.

Would that work?

2. The existing Maya fallback Transform3d handler derives from UsdTransform3dMayaXformStack, and doesn't define any undoable commands: it just uses the ones from its parent class. 
https://github.com/Autodesk/maya-usd/blob/395ad3177179161378296d85a822a63b5dcc9366/lib/mayaUsd/ufe/UsdTransform3dFallbackMayaXformStack.h

Is there any way to re-use this strategy in your own code, since you're also deriving from UsdTransform3dMayaXformStack?

**Reply by csyshing** [reply]
Thanks @ppt-adsk !
After discussed with our team internally, we have decided to use the functionality as it is from maya-usd, instead of adding custom handler.

**Reply by ppt-adsk** [reply]
Hey @csyshing, good to hear.  Sometimes the best option is the simplest one, if you can manage the compromises.
